include/libhal-soft/i2c_bit_bang.hpp

```diff
@@ -0,0 +1,210 @@
+                  function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief this function will write a single bit at a time, dealing with
+   * simulating the clock and the clock stretching feature
+   *
+   * @param p_bit_to_write the bit which will be written on the bus
+   * @param p_timeout a timeout function which is primarily used for clock
+   * stretching to ensure the peripheral doesn't hold the line too long
+   */
+  void write_bit(hal::byte p_bit_to_write,
+                 function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief this function will read in as many bytes as allocated inside of the
+   * span while also acking or nacking the data
+   *
+   * @param p_data_in a span which will be filled with the bytes that will be
+   * read from the bus
+   * @param p_timeout a timeout function which is primarily used for clock
+   * stretching to ensure the peripheral doesn't hold the line too long
+   */
+  void read(std::span<hal::byte> p_data_in,
+            function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief this function is responsible for reading a byte at a time off the
+   * bus and also creating the byte from bits
+   *
+   * @return hal::byte the byte that has been read off of the bus
+   */
+  hal::byte read_byte();
+
+  /**
+   * @brief this function is responsible for reading a single bit at a time off
+   * of the bus while also managing the clock line. It will release (pull sda
+   * high) every time it is called
+   *
+   * @return hal::byte
+   */
+  hal::byte read_bit();
+
+  /**
+   * @brief this function is a high speed version of the hal::delay function
+   * which operates on ticks rather then a time duration where hal::delay
+   * provides delays up to 17us, this function will provide delays up to
+   * about 2.7us making it optimal for high accuracy delay
+   *
+   * @param ticks the amount of ticks this function will delay for, this should
+   * be based on the frequency of the clock passed into the bit_bang class
+   */
+  [[gnu::always_inline]] inline void high_speed_delay(uint64_t ticks);
+
+  /**
+   * @brief an output pin which will mimic the behavior of an I2C scl pin
+   */
+  output_pin* m_scl;
+  /**
+   * @brief an output pin which will mimic the bahavior of an I2C sda pin
+   */
+  output_pin* m_sda;
+  /**
+   * @brief a steady_clock which will be used to correctly time the different
+   * frequencies
+   */
+  steady_clock* m_clock;
+
+  /**
+   * @brief the time that scl will be held high for
+   */
+  uint64_t m_scl_high_ticks;
+  /**
+   * @brief the time that scl will be held low for
+   */
+  uint64_t m_scl_low_ticks;
+
+  /**
+   * @brief the amount of ticks that the uptime function takes to run, we store
+   * this so we save computation time in our high speed delay function
+   */
+  uint32_t m_uptime_ticks;
+
+  /**
+   * @brief all the information that the bus will need to operate on
+   */
+  bus_info m_bus;
```

Comment:
```suggestion
  float m_duty_cycle;
```

If we change bus_info the class's structure changes and this becomes an ABI break. You can simply store the parts you actually need.